cpp/include/raft/core/mdbuffer.cuh

```diff
@@ -0,0 +1,831 @@
+                              typename container_policy_type::template container_policy<MemType>>;
+  // We use the static cast here to ensure that the memory types appear in the
+  // order expected for retrieving the correct variant alternative based on
+  // memory type. Even if the memory types are re-arranged in the enum and
+  // assigned different values, the logic should remain correct.
+  using owning_type_variant = std::variant<owning_type<static_cast<memory_type>(0)>,
+                                           owning_type<static_cast<memory_type>(1)>,
+                                           owning_type<static_cast<memory_type>(2)>,
+                                           owning_type<static_cast<memory_type>(3)>>;
+
+  template <memory_type MemType>
+  using view_type = typename owning_type<MemType>::view_type;
+
+  using view_type_variant = std::variant<view_type<static_cast<memory_type>(0)>,
+                                         view_type<static_cast<memory_type>(1)>,
+                                         view_type<static_cast<memory_type>(2)>,
+                                         view_type<static_cast<memory_type>(3)>>;
+
+  template <memory_type MemType>
+  using const_view_type         = typename owning_type<MemType>::const_view_type;
+  using const_view_type_variant = std::variant<const_view_type<static_cast<memory_type>(0)>,
+                                               const_view_type<static_cast<memory_type>(1)>,
+                                               const_view_type<static_cast<memory_type>(2)>,
+                                               const_view_type<static_cast<memory_type>(3)>>;
+
+  using storage_type_variant = concatenated_variant_t<view_type_variant, owning_type_variant>;
+
+  template <memory_type MemType, bool is_owning>
+  using storage_type =
+    std::variant_alternative_t<std::size_t{is_owning} * std::variant_size_v<view_type_variant> +
+                                 std::size_t{variant_index_from_memory_type(MemType)},
+                               storage_type_variant>;
+
+  /**
+   * @brief Construct an empty, uninitialized buffer
+   */
+  constexpr mdbuffer() = default;
+
+ private:
+  container_policy_type cp_{};
+  storage_type_variant data_{};
+
+  template <typename FromT, std::size_t FromIndex, std::size_t ToIndex>
+  auto static constexpr is_copyable_combination()
+  {
+    return detail::mdspan_copyable_v<
+      decltype(std::declval<std::variant_alternative_t<ToIndex, owning_type_variant>>().view()),
+      std::variant_alternative_t<FromIndex, decltype(std::declval<FromT>().view())>>;
+  }
+
+  template <std::size_t FromIndex, typename FromT, std::size_t... Is>
+  auto static constexpr get_copyable_combinations(std::index_sequence<Is...>)
+  {
+    return std::array{is_copyable_combination<FromT, FromIndex, Is>()...};
+  }
+
+  template <typename FromT, std::size_t... Is>
+  auto static constexpr get_copyable_combinations(bool, std::index_sequence<Is...>)
```

Comment:
What is `bool` parameter here for?

Reply:
Just added a comment in the code because it took me a minute for me to figure that out for myself.